### ISSUE TITLE
Cuda allocator pointer alignment

### ISSUE THREAD

**Issue by skakunm** [microsoft/onnxruntime#12289]
**Describe the bug**
Cuda allocator returns pointers that are not aligned correctly. When I read cudaDeviceProp::textureAlignment, the value is 512, but Ort::Allocator::Alloc() returns pointers aligned to 256. I initialize Ort::Allocator using Ort::MemoryInfo("Cuda", OrtAllocatorType::OrtArenaAllocator, deviceIndex, OrtMemTypeDefault), but I've tried OrtAllocatorType::OrtDefaultAllocator and it does not change the alignment. If I allocate some memory using cudaMalloc I always get pointers aligned to 512. 

This alignment issue prevents me from creating a Cuda texture because pointer must be aligned properly. If I just run kernels on that memory, it runs fine, just texture creation is a problem.

**Urgency**
None

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- ONNX Runtime installed from (source or binary): binary
- ONNX Runtime version: 1.10
- Python version: 3.8
- Visual Studio version (if applicable): 2019
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.2
- GPU model and memory: 2080 Ti, 11GB

**To Reproduce**
- Create Ort::MemoryInfo object with arguments ("Cuda", OrtAllocatorType::OrtArenaAllocator, deviceIndex, OrtMemTypeDefault)
- Create Ort::Allocator object with session with these provider options changed:
            OrtCUDAProviderOptions providerOptions;
            providerOptions.device_id = DeviceIndex;
            providerOptions.has_user_compute_stream = 1;
            providerOptions.user_compute_stream = 0;
- Run Ort::Allocator with 1048576 bytes
- I get pointer aligned to 256
- When I call cudaTextureCreate it sometimes fails because pointer is not aligned to 512
- I have tried this with many different models. Behavior is that for some it will be well aligned, by chance, and for some it won't. If I add a few more dummy allocations, it eventually does align with 512 and the texture creation works.

**Expected behavior**
I expect Ort::Allocator to provide arrays that are well aligned, so that output can be used to create a Cuda texture.

**Screenshots**
None

**Additional context**
None

**Comment by zhanghuanrong** [comment]
Thanks for reporting this issue.  This is an interesting scenario.
I suppose you are just requesting the output tensor could be 512 aligned, not all. correct?
We will investigate this.

**Comment by hariharans29** [comment]
The alignment being 256 is due to the way the Cuda allocator is implemented (an internal detail) and as such we haven't exposed an interface that takes in the alignment requirement from the user.

Any reason you are limited to using the OrtAllocator to allocate Cuda textures in your app or could you just use your own allocator ?

**Comment by skakunm** [comment]
Yes, I would need only the output to be alligned to 512 (I guess this is true for my gpu, and can possibly change from gpu to gpu, but I'm not sure). I guess it would also make sense to allign the input. I can't come up with some exact scenario, but since the user can use it, someone might want to use it to create a texture at some point.

I guessed there is something else going on, other than just cudaMalloc, when calling the Ort::Allocator. Documentation is not exactly detailed regarding this. But if that is not the case I can allocate it myself using standard cudaMalloc. That sounds like a simplest workaround, other workaround I found is creating another array and copying data to it, since copy works regardles og the alignment. But I don't like adding extra unnecessary operations.

Anyway, I really feel like either documentation should be more explicit regarding this, or it should be properly aligned by default (second seems much better). When it does happen, it is a hard bug to find.
